app/controllers/provider/admin/keys_controller.rb

```diff
@@ -26,8 +28,9 @@ def create
 
     if @key.persisted?
       @keys = @cinstance.application_keys.pluck_values
+      @notice = t('.create.success')
     else
-      flash.now[:error] = "Invalid key. Please review and try submitting again."
+      @cinstance.errors.add(:keys, :invalid)
```

Comment:
I think this is not ideal, because it doesn't give enough info...
The actual error would be coming from `@key.errors`, which could include:

`"Value has already been taken"` or `"Limit reached"` (as only 5 keys are allowed).

However, value can also be e.g. `["Value is too short (minimum is 5 characters)"]`, which would be covered by the current `cinstance.keys.invalid` message.

Reply:
maybe merge errors then? I believe I've seen this done somewhere else.

Reply:
> I believe I've seen this done somewhere else.

yeah, that rings a bell, but I don't remember where I saw it either :sweat_smile:

Reply:
I'll try to improve it but I would want to block this PR on something that's been in the code since forever. Another PR for improving error handling would probably better.